In [1]:
import pandas as pd

In [2]:
na_vals = ['NA', 'Missing']
df = pd.read_csv('data/survey_results_public.csv', index_col = "Respondent")
schema_df = pd.read_csv('data/survey_results_schema.csv', index_col = 'Column')

In [3]:
df

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   
...                                                       ...      ...   
88377                                                     NaN      Yes   
88601                                                     NaN       No   
88802                                                     NaN       No   
88816                                                     NaN       No   
88863                                                     NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
1                                                       Never   
2                                     Less than once per year   
3                                                       Never   
4                                                       Never   
5                                  Once a month or more often   
...                                                       ...   
88377       Less than once a month but more than once per ...   
88601                                                   Never   
88802                                                   Never   
88816                                                   Never   
88863                                 Less than once per year   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   
...                                                       ...   
88377       The quality of OSS and closed source software ...   
88601       The quality of OSS and closed source software ...   
88802                                                     NaN   
88816       OSS is, on average, of HIGHER quality than pro...   
88863       OSS is, on average, of HIGHER quality than pro...   

                                                   Employment  \
Respondent                                                      
1                      Not employed, and not looking for work   
2                          Not employed, but looking for work   
3                                          Employed full-time   
4                                          Employed full-time   
5                                          Employed full-time   
...                                                       ...   
88377                  Not employed, and not looking for work   
88601                                                     NaN   
88802                                      Employed full-time   
88816       Independent contractor, freelancer, or self-em...   
88863                  Not employed, and not looking for work   

                           Country         Student  \
Respondent                                           
1                   United Kingdom              No   
2           Bosnia and Herzegovina  Yes, full-time   
3                         Thailand              No   
4                    United States              No   
5                          Ukraine              No   
...                            ...             ...   
88377                       Canada              No   
88601                          NaN             NaN   
88802     

In [4]:
pd.set_option('display.max_columns',85) 
pd.set_option('display.max_rows',85) 

In [5]:
schema_df

QuestionText
Column                                                                   
Respondent              Randomized respondent ID number (not in order ...
MainBranch              Which of the following options best describes ...
Hobbyist                                          Do you code as a hobby?
OpenSourcer                   How often do you contribute to open source?
OpenSource              How do you feel about the quality of open sour...
Employment              Which of the following best describes your cur...
Country                         In which country do you currently reside?
Student                 Are you currently enrolled in a formal, degree...
EdLevel                 Which of the following best describes the high...
UndergradMajor          What was your main or most important field of ...
EduOther                Which of the following types of non-degree edu...
OrgSize                 Approximately how many people are employed by ...
DevType                 Which of the following describe you? Please se...
YearsCode               Including any education, how many years have y...
Age1stCode              At what age did you write your first line of c...
YearsCodePro            How many years have you coded professionally (...
CareerSat               Overall, how satisfied are you with your caree...
JobSat                  How satisfied are you with your current job? (...
MgrIdiot                How confident are you that your manager knows ...
MgrMoney                Do you believe that you need to be a manager t...
MgrWant                 Do you want to become a manager yourself in th...
JobSeek                 Which of the following best describes your cur...
LastHireDate            When was the last time that you took a job wit...
LastInt                 In your most recent successful job interview (...
FizzBuzz                Have you ever been asked to solve FizzBuzz in ...
JobFactors              Imagine that you are deciding between two job ...
ResumeUpdate            Think back to the last time you updated your r...
CurrencySymbol          Which currency do you use day-to-day? If your ...
CurrencyDesc            Which currency do you use day-to-day? If your ...
CompTotal               What is your current total compensation (salar...
CompFreq                 Is that compensation weekly, monthly, or yearly?
ConvertedComp           Salary converted to annual USD salaries using ...
WorkWeekHrs              On average, how many hours per week do you work?
WorkPlan                          How structured or planned is your work?
WorkChallenge           Of these options, what are your greatest chall...
WorkRemote                                How often do you work remotely?
WorkLoc                                   Where would you prefer to work?
ImpSyn                  For the specific work you do, and the years of...
CodeRev                          Do you review code as part of your work?
CodeRevHrs              On average, how many hours per week do you spe...
UnitTests               Does your company regularly employ unit tests ...
PurchaseHow             How does your company make decisions about pur...
PurchaseWhat            What level of influence do you, personally, ha...
LanguageWorkedWith      Which of the following programming, scripting,...
LanguageDesireNextYear  Which of the following programming, scripting,...
DatabaseWorkedWith      Which of the following database environments h...
DatabaseDesireNextYear  Which of the following database environments h...
PlatformWorkedWith      Which of the following platforms have you done...
PlatformDesireNextYear  Which of the following platforms have you done...
WebFrameWorkedWith      Which of the following web frameworks have you...
WebFrameDesireNextYear  Which of the following web frameworks have you...
MiscTechWorkedWith      Which of the following other frameworks, libra...
MiscTechDesireNextYear  Which of the following other frameworks, 

In [6]:
#This sorts the columns in Descending order
schema_df.sort_index()

QuestionText
Column                                                                   
Age                     What is your age (in years)? If you prefer not...
Age1stCode              At what age did you write your first line of c...
BetterLife              Do you think people born today will have a bet...
BlockchainIs            Blockchain / cryptocurrency technology is prim...
BlockchainOrg           How is your organization thinking about or imp...
CareerSat               Overall, how satisfied are you with your caree...
CodeRev                          Do you review code as part of your work?
CodeRevHrs              On average, how many hours per week do you spe...
CompFreq                 Is that compensation weekly, monthly, or yearly?
CompTotal               What is your current total compensation (salar...
Containers              How do you use containers (Docker, Open Contai...
ConvertedComp           Salary converted to annual USD salaries using ...
Country                         In which country do you currently reside?
CurrencyDesc            Which currency do you use day-to-day? If your ...
CurrencySymbol          Which currency do you use day-to-day? If your ...
DatabaseDesireNextYear  Which of the following database environments h...
DatabaseWorkedWith      Which of the following database environments h...
Dependents              Do you have any dependents (e.g., children, el...
DevEnviron              Which development environment(s) do you use re...
DevType                 Which of the following describe you? Please se...
EdLevel                 Which of the following best describes the high...
EduOther                Which of the following types of non-degree edu...
Employment              Which of the following best describes your cur...
EntTeams                Have you ever used Stack Overflow for Enterpri...
Ethnicity               Which of the following do you identify as? Ple...
Extraversion              Do you prefer online chat or IRL conversations?
FizzBuzz                Have you ever been asked to solve FizzBuzz in ...
Gender                  Which of the following do you currently identi...
Hobbyist                                          Do you code as a hobby?
ITperson                 Are you the "IT support person" for your family?
ImpSyn                  For the specific work you do, and the years of...
JobFactors              Imagine that you are deciding between two job ...
JobSat                  How satisfied are you with your current job? (...
JobSeek                 Which of the following best describes your cur...
LanguageDesireNextYear  Which of the following programming, scripting,...
LanguageWorkedWith      Which of the following programming, scripting,...
LastHireDate            When was the last time that you took a job wit...
LastInt                 In your most recent successful job interview (...
MainBranch              Which of the following options best describes ...
MgrIdiot                How confident are you that your manager knows ...
MgrMoney                Do you believe that you need to be a manager t...
MgrWant                 Do you want to become a manager yourself in th...
MiscTechDesireNextYear  Which of the following other frameworks, libra...
MiscTechWorkedWith      Which of the following other frameworks, libra...
OffOn                         Have you tried turning it off and on again?
OpSys                   What is the primary operating system in which ...
OpenSource              How do you feel about the quality of open sour...
OpenSourcer                   How often do you contribute to open source?
OrgSize                 Approximately how many people are employed by ...
PlatformDesireNextYear  Which of the following platforms have you done...
PlatformWorkedWith      Which of the following platforms have you done...
PurchaseHow             How does your company make decisions about pur...
PurchaseWhat            What level of influence do you, personall

In [7]:
df.head()

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

In [8]:
df.shape

(88883, 84)

In [9]:
df.columns

Index(['MainBranch', 'Hobbyist', 'OpenSourcer', 'OpenSource', 'Employment',
       'Country', 'Student', 'EdLevel', 'UndergradMajor', 'EduOther',
       'OrgSize', 'DevType', 'YearsCode', 'Age1stCode', 'YearsCodePro',
       'CareerSat', 'JobSat', 'MgrIdiot', 'MgrMoney', 'MgrWant', 'JobSeek',
       'LastHireDate', 'LastInt', 'FizzBuzz', 'JobFactors', 'ResumeUpdate',
       'CurrencySymbol', 'CurrencyDesc', 'CompTotal', 'CompFreq',
       'ConvertedComp', 'WorkWeekHrs', 'WorkPlan', 'WorkChallenge',
       'WorkRemote', 'WorkLoc', 'ImpSyn', 'CodeRev', 'CodeRevHrs', 'UnitTests',
       'PurchaseHow', 'PurchaseWhat', 'LanguageWorkedWith',
       'LanguageDesireNextYear', 'DatabaseWorkedWith',
       'DatabaseDesireNextYear', 'PlatformWorkedWith',
       'PlatformDesireNextYear', 'WebFrameWorkedWith',
       'WebFrameDesireNextYear', 'MiscTechWorkedWith',
       'MiscTechDesireNextYear', 'DevEnviron', 'OpSys', 'Containers',
       'BlockchainOrg', 'BlockchainIs', 'BetterLife', 'ITperson', 

In [10]:
df['Hobbyist']

Respondent
1        Yes
2         No
3        Yes
4         No
5        Yes
        ... 
88377    Yes
88601     No
88802     No
88816     No
88863    Yes
Name: Hobbyist, Length: 88883, dtype: object

In [11]:
df.loc[1]

MainBranch                           I am a student who is learning to code
Hobbyist                                                                Yes
OpenSourcer                                                           Never
OpenSource                The quality of OSS and closed source software ...
Employment                           Not employed, and not looking for work
Country                                                      United Kingdom
Student                                                                  No
EdLevel                                           Primary/elementary school
UndergradMajor                                                          NaN
EduOther                  Taught yourself a new language, framework, or ...
OrgSize                                                                 NaN
DevType                                                                 NaN
YearsCode                                                                 4
Age1stCode  

all the values of the first row

In [12]:
df.loc[1,'Hobbyist']

'Yes'

row 0 of Hobbyist column

In [13]:
df.loc[[1,2,3],'Hobbyist']

Respondent
1    Yes
2     No
3    Yes
Name: Hobbyist, dtype: object

row 1,2,3 of column Hobbyist

In [14]:
df.loc[:3, 'Hobbyist']

Respondent
1    Yes
2     No
3    Yes
Name: Hobbyist, dtype: object

In [15]:
df.loc[:3, 'Hobbyist':'Employment']

Hobbyist              OpenSourcer  \
Respondent                                     
1               Yes                    Never   
2                No  Less than once per year   
3               Yes                    Never   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   

                                        Employment  
Respondent                                          
1           Not employed, and not looking for work  
2               Not employed, but looking for work  
3                               Employed full-time

first 3 rows from column Hobbyist through Employment(inclusive)

In [16]:
df.head()

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

we can see that there is already a unique identifier as respondent column, so we do not need the default index, hence lets delete it. There are 2 ways to do it, either reset the index or by df.reset_index() or just go uo where we are reading the file and add an attribute as index_col="Respondent"

In [17]:
schema_df.loc['Hobbyist']

QuestionText    Do you code as a hobby?
Name: Hobbyist, dtype: object

In [18]:
schema_df.loc['MgrIdiot', 'QuestionText']

'How confident are you that your manager knows what they’re doing?'

In [19]:
high_salary = (df['ConvertedComp'] > 70000)

In [20]:
df.loc[high_salary]

MainBranch Hobbyist  \
Respondent                                                               
6           I am not primarily a developer, but I write co...      Yes   
9                              I am a developer by profession      Yes   
13                             I am a developer by profession      Yes   
16                             I am a developer by profession      Yes   
22                             I am a developer by profession      Yes   
...                                                       ...      ...   
88876                          I am a developer by profession      Yes   
88877                          I am a developer by profession      Yes   
88878                          I am a developer by profession      Yes   
88879                          I am a developer by profession      Yes   
88882                          I am a developer by profession      Yes   

                                                  OpenSourcer  \
Respondent                                                      
6                                                       Never   
9                                  Once a month or more often   
13          Less than once a month but more than once per ...   
16                                                      Never   
22                                    Less than once per year   
...                                                       ...   
88876                                                   Never   
88877                                 Less than once per year   
88878                                 Less than once per year   
88879                                                   Never   
88882                                                   Never   

                                                   OpenSource  \
Respondent                                                      
6           The quality of OSS and closed source software ...   
9           The quality of OSS and closed source software ...   
13          OSS is, on average, of HIGHER quality than pro...   
16          The quality of OSS and closed source software ...   
22          OSS is, on average, of HIGHER quality than pro...   
...                                                       ...   
88876       OSS is, on average, of HIGHER quality than pro...   
88877       OSS is, on average, of HIGHER quality than pro...   
88878       The quality of OSS and closed source software ...   
88879       The quality of OSS and closed source software ...   
88882       OSS is, on average, of LOWER quality than prop...   

                    Employment         Country         Student  \
Respondent                                                       
6           Employed full-time          Canada              No   
9           Employed full-time     New Zealand              No   
13          Employed full-time   United States              No   
16          Employed full-time  United Kingdom              No   
22          Employed full-time   United States              No   
...                        ...             ...             ...   
88876       Employed full-time   United States              No   
88877       Employed full-time   United States              No   
88878       Employed full-time   United States              No   
88879       Employed full-time         Finland              No   
88882       Employed full-time     Netherlands  Yes, full-time   

                                                      EdLevel  \
Respondent                                                      
6                    Bachelor’s degree (BA, BS, B.Eng., etc.)   
9           Some college/university study without earning ...   
13                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
16                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
22          Some college/university study without earning ...   
...                                                       ...   
88876            

In [21]:
df.loc[high_salary, ['Country', 'LanguageWorkedWith', 'ConvertedComp']]

Country                                 LanguageWorkedWith  \
Respondent                                                                      
6                   Canada                                         Java;R;SQL   
9              New Zealand  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
13           United States  Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...   
16          United Kingdom  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;T...   
22           United States  Bash/Shell/PowerShell;C++;HTML/CSS;JavaScript;...   
...                    ...                                                ...   
88876        United States  Bash/Shell/PowerShell;C#;HTML/CSS;Java;Python;...   
88877        United States  Bash/Shell/PowerShell;C;Clojure;HTML/CSS;Java;...   
88878        United States               HTML/CSS;JavaScript;Scala;TypeScript   
88879              Finland                   Bash/Shell/PowerShell;C++;Python   
88882          Netherlands             C#;HTML/CSS;Java;JavaScript;PHP;Python   

            ConvertedComp  
Respondent                 
6                366420.0  
9                 95179.0  
13                90000.0  
16               455352.0  
22               103000.0  
...                   ...  
88876            180000.0  
88877           2000000.0  
88878            130000.0  
88879             82488.0  
88882            588012.0  

[22289 rows x 3 columns]

Let's filter it on countries, for that we'll create a list of countries instead of making filter like country == '' | country == '' , this will be very tedious.

In [22]:
countries = ['United States', 'India', 'United Kingdom', 'Germany', 'Canada']
filt1 = df['Country'].isin(countries)

In [23]:
df.loc[filt1, 'Country']

Respondent
1        United Kingdom
4         United States
6                Canada
8                 India
10                India
              ...      
85642     United States
85961    United Kingdom
86012             India
88282     United States
88377            Canada
Name: Country, Length: 45008, dtype: object

lets filter out only those people who know python

In [24]:
filt2 = df['LanguageWorkedWith'].str.contains('Python', na = False)

In [25]:
filt2

Respondent
1         True
2         True
3        False
4         True
5         True
         ...  
88377    False
88601    False
88802    False
88816    False
88863    False
Name: LanguageWorkedWith, Length: 88883, dtype: bool

In [26]:
df.loc[filt2, "LanguageWorkedWith"]

Respondent
1                          HTML/CSS;Java;JavaScript;Python
2                                      C++;HTML/CSS;Python
4                                      C;C++;C#;Python;SQL
5              C++;HTML/CSS;Java;JavaScript;Python;SQL;VBA
8        Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
                               ...                        
84539    Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
85738      Bash/Shell/PowerShell;C++;Python;Ruby;Other(s):
86566      Bash/Shell/PowerShell;HTML/CSS;Python;Other(s):
87739             C;C++;HTML/CSS;JavaScript;PHP;Python;SQL
88212                           HTML/CSS;JavaScript;Python
Name: LanguageWorkedWith, Length: 36443, dtype: object

In [27]:
df.rename(columns = {'ConvertedComp': 'SalaryUSD'}, inplace = True)

In [28]:
df['SalaryUSD']

Respondent
1            NaN
2            NaN
3         8820.0
4        61000.0
5            NaN
          ...   
88377        NaN
88601        NaN
88802        NaN
88816        NaN
88863        NaN
Name: SalaryUSD, Length: 88883, dtype: float64

In [29]:
df['Hobbyist']

Respondent
1        Yes
2         No
3        Yes
4         No
5        Yes
        ... 
88377    Yes
88601     No
88802     No
88816     No
88863    Yes
Name: Hobbyist, Length: 88883, dtype: object

In [30]:
df['Hobbyist'].map({'Yes': True, 'No': False})

Respondent
1         True
2        False
3         True
4        False
5         True
         ...  
88377     True
88601    False
88802    False
88816    False
88863     True
Name: Hobbyist, Length: 88883, dtype: bool

In [31]:
df['Hobbyist'] = df['Hobbyist'].map({'Yes': True, 'No': False})

In [32]:
df

MainBranch  Hobbyist  \
Respondent                                                                
1                      I am a student who is learning to code      True   
2                      I am a student who is learning to code     False   
3           I am not primarily a developer, but I write co...      True   
4                              I am a developer by profession     False   
5                              I am a developer by profession      True   
...                                                       ...       ...   
88377                                                     NaN      True   
88601                                                     NaN     False   
88802                                                     NaN     False   
88816                                                     NaN     False   
88863                                                     NaN      True   

                                                  OpenSourcer  \
Respondent                                                      
1                                                       Never   
2                                     Less than once per year   
3                                                       Never   
4                                                       Never   
5                                  Once a month or more often   
...                                                       ...   
88377       Less than once a month but more than once per ...   
88601                                                   Never   
88802                                                   Never   
88816                                                   Never   
88863                                 Less than once per year   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   
...                                                       ...   
88377       The quality of OSS and closed source software ...   
88601       The quality of OSS and closed source software ...   
88802                                                     NaN   
88816       OSS is, on average, of HIGHER quality than pro...   
88863       OSS is, on average, of HIGHER quality than pro...   

                                                   Employment  \
Respondent                                                      
1                      Not employed, and not looking for work   
2                          Not employed, but looking for work   
3                                          Employed full-time   
4                                          Employed full-time   
5                                          Employed full-time   
...                                                       ...   
88377                  Not employed, and not looking for work   
88601                                                     NaN   
88802                                      Employed full-time   
88816       Independent contractor, freelancer, or self-em...   
88863                  Not employed, and not looking for work   

                           Country         Student  \
Respondent                                           
1                   United Kingdom              No   
2           Bosnia and Herzegovina  Yes, full-time   
3                         Thailand              No   
4                    United States              No   
5                          Ukraine              No   
...                            ...             ...   
88377                       Canada              No   
88601                          NaN             NaN 

In [33]:
#lets convert back the Hobbyist to yes or No from True or False
df['Hobbyist'] = df['Hobbyist'].map({True:'Yes', False: 'No'})
df['Hobbyist']

Respondent
1        Yes
2         No
3        Yes
4         No
5        Yes
        ... 
88377    Yes
88601     No
88802     No
88816     No
88863    Yes
Name: Hobbyist, Length: 88883, dtype: object

In [34]:
df.sort_values(by = ['Country','SalaryUSD'], ascending =[True, False])


MainBranch Hobbyist  \
Respondent                                                               
63129                          I am a developer by profession      Yes   
50499       I am not primarily a developer, but I write co...      Yes   
39258                          I am a developer by profession      Yes   
58450                          I am a developer by profession      Yes   
7085                           I am a developer by profession      Yes   
...                                                       ...      ...   
88062                                                     NaN       No   
88076                                                     NaN       No   
88601                                                     NaN       No   
88802                                                     NaN       No   
88816                                                     NaN       No   

                        OpenSourcer  \
Respondent                            
63129                         Never   
50499       Less than once per year   
39258       Less than once per year   
58450                         Never   
7085        Less than once per year   
...                             ...   
88062                         Never   
88076                         Never   
88601                         Never   
88802                         Never   
88816                         Never   

                                                   OpenSource  \
Respondent                                                      
63129       OSS is, on average, of HIGHER quality than pro...   
50499       OSS is, on average, of LOWER quality than prop...   
39258       OSS is, on average, of LOWER quality than prop...   
58450       OSS is, on average, of HIGHER quality than pro...   
7085        The quality of OSS and closed source software ...   
...                                                       ...   
88062       OSS is, on average, of LOWER quality than prop...   
88076       OSS is, on average, of HIGHER quality than pro...   
88601       The quality of OSS and closed source software ...   
88802                                                     NaN   
88816       OSS is, on average, of HIGHER quality than pro...   

                                                   Employment      Country  \
Respondent                                                                   
63129                                      Employed full-time  Afghanistan   
50499                                      Employed full-time  Afghanistan   
39258                                      Employed full-time  Afghanistan   
58450                                      Employed full-time  Afghanistan   
7085        Independent contractor, freelancer, or self-em...  Afghanistan   
...                                                       ...          ...   
88062                                                     NaN          NaN   
88076                                      Employed full-time          NaN   
88601                                                     NaN          NaN   
88802                                      Employed full-time          NaN   
88816       Independent contractor, freelancer, or self-em...          NaN   

                   Student                                   EdLevel  \
Respondent                                                             
63129       Yes, full-time    I never completed any formal education   
50499                   No  Bachelor’s degree (BA, BS, B.Eng., etc.)   
39258                   No  Bachelor’s degree (BA, BS, B.Eng., etc.)   
58450                   No  Bachelor’s degree (BA, BS, B.Eng., etc.)   
7085                    No  Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                    ...                                       ...   
88062                  NaN                                       NaN   
88076                  NaN                                       NaN   

In [35]:
df[['Country','SalaryUSD']].head(50)

Country  SalaryUSD
Respondent                                   
1                   United Kingdom        NaN
2           Bosnia and Herzegovina        NaN
3                         Thailand     8820.0
4                    United States    61000.0
5                          Ukraine        NaN
6                           Canada   366420.0
7                          Ukraine        NaN
8                            India        NaN
9                      New Zealand    95179.0
10                           India    13293.0
11             Antigua and Barbuda        NaN
12                          Canada        NaN
13                   United States    90000.0
14                         Germany    57060.0
15                           India        NaN
16                  United Kingdom   455352.0
17                       Australia    65277.0
18              Russian Federation    21996.0
19                          Brazil    31140.0
20                       Lithuania    41244.0
21                          Canada        NaN
22                   United States   103000.0
23                   United States    69000.0
24                          Israel    26388.0
25                    South Africa     6300.0
26                   United States   114000.0
27                        Colombia        NaN
28                          Turkey        NaN
29                   United States   100000.0
30                     Switzerland    90417.0
31                          Canada        NaN
32                   United States  1100000.0
33                       Argentina     8400.0
34                       Sri Lanka        NaN
35                   United States    80000.0
36                         Germany    20628.0
37                          Canada        NaN
38                  Czech Republic    51312.0
39                   United States   156000.0
40                         Denmark        NaN
41                        Malaysia        NaN
42                      Bangladesh        NaN
43                           Spain    20623.0
44                         Germany    87996.0
45                   United States   175000.0
46                   United States        NaN
47                   United States   187000.0
48              Russian Federation        NaN
49                          Serbia    60000.0
50                           India     5597.0

In [36]:
#10 largest salaries
df['SalaryUSD'].nlargest(10)

Respondent
58      2000000.0
102     2000000.0
166     2000000.0
436     2000000.0
452     2000000.0
491     2000000.0
539     2000000.0
770     2000000.0
789     2000000.0
1232    2000000.0
Name: SalaryUSD, dtype: float64

In [37]:
#all the details of people making 10 largest salaries
df.nlargest(10,'SalaryUSD')

MainBranch Hobbyist  \
Respondent                                            
58          I am a developer by profession      Yes   
102         I am a developer by profession       No   
166         I am a developer by profession      Yes   
436         I am a developer by profession      Yes   
452         I am a developer by profession      Yes   
491         I am a developer by profession      Yes   
539         I am a developer by profession       No   
770         I am a developer by profession      Yes   
789         I am a developer by profession      Yes   
1232        I am a developer by profession       No   

                                                  OpenSourcer  \
Respondent                                                      
58                                 Once a month or more often   
102                                                     Never   
166                                   Less than once per year   
436                                   Less than once per year   
452                                                     Never   
491                                   Less than once per year   
539         Less than once a month but more than once per ...   
770                                   Less than once per year   
789                                   Less than once per year   
1232                                                    Never   

                                                   OpenSource  \
Respondent                                                      
58          OSS is, on average, of LOWER quality than prop...   
102         The quality of OSS and closed source software ...   
166         The quality of OSS and closed source software ...   
436         The quality of OSS and closed source software ...   
452         The quality of OSS and closed source software ...   
491                                                       NaN   
539         OSS is, on average, of HIGHER quality than pro...   
770         The quality of OSS and closed source software ...   
789         The quality of OSS and closed source software ...   
1232        OSS is, on average, of HIGHER quality than pro...   

                    Employment        Country         Student  \
Respondent                                                      
58          Employed full-time  United States              No   
102         Employed full-time  United States              No   
166         Employed full-time  United States              No   
436         Employed full-time  United States              No   
452         Employed full-time  United States  Yes, full-time   
491         Employed full-time  United States              No   
539         Employed full-time  United States              No   
770         Employed full-time  United States              No   
789         Employed full-time  United States  Yes, full-time   
1232        Employed full-time  United States              No   

                                                      EdLevel  \
Respondent                                                      
58                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
102                  Bachelor’s degree (BA, BS, B.Eng., etc.)   
166                  Bachelor’s degree (BA, BS, B.Eng., etc.)   
436                  Bachelor’s degree (BA, BS, B.Eng., etc.)   
452                    I never completed any formal education   
491                  Bachelor’s degree (BA, BS, B.Eng., etc.)   
539         Some college/university study without earning ...   
770               Master’s degree (MA, MS, M.Eng., MBA, etc.)   
789               Master’s degree (MA, MS, M.Eng., MBA, etc.)   
1232                 Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
58          Computer science, computer engineering, or sof...   
102         Computer science, computer engineering, 

In [38]:
df['SalaryUSD'].nsmallest(10)

Respondent
280     0.0
293     0.0
722     0.0
1105    0.0
1501    0.0
1685    0.0
1782    0.0
2019    0.0
2095    0.0
2665    0.0
Name: SalaryUSD, dtype: float64

### Aggregate Functions

In [39]:
df['SalaryUSD'].head(15)

Respondent
1          NaN
2          NaN
3       8820.0
4      61000.0
5          NaN
6     366420.0
7          NaN
8          NaN
9      95179.0
10     13293.0
11         NaN
12         NaN
13     90000.0
14     57060.0
15         NaN
Name: SalaryUSD, dtype: float64

In [40]:
df['SalaryUSD'].median()

57287.0

In [41]:
df.median()

TypeError: could not convert string to float: 'I am a student who is learning to code'

We are having trouble in getting median values since there are string dtypes in our dataframe, hence lets ask to give median of only float dtypes.

In [42]:
import numpy as np
df.select_dtypes(include=[np.number]).median()

CompTotal      62000.0
SalaryUSD      57287.0
WorkWeekHrs       40.0
CodeRevHrs         4.0
Age               29.0
dtype: float64

why not use mean i.e. average instead of median?
Since there are few very high salaires, they contribute heavily in exploding the average hence we dont get skewed idea of salaries, so its better to use median. 

In [43]:
#A quick overview of data
df.describe()

CompTotal     SalaryUSD   WorkWeekHrs    CodeRevHrs           Age
count  5.594500e+04  5.582300e+04  64503.000000  49790.000000  79210.000000
mean   5.519014e+11  1.271107e+05     42.127197      5.084308     30.336699
std    7.331926e+13  2.841523e+05     37.287610      5.513931      9.178390
min    0.000000e+00  0.000000e+00      1.000000      0.000000      1.000000
25%    2.000000e+04  2.577750e+04     40.000000      2.000000     24.000000
50%    6.200000e+04  5.728700e+04     40.000000      4.000000     29.000000
75%    1.200000e+05  1.000000e+05     44.750000      6.000000     35.000000
max    1.000000e+16  2.000000e+06   4850.000000     99.000000     99.000000

In [44]:
df['SalaryUSD'].count()

55823

there are total around 80000 people but in salaryusd count we're getting only 55823 which means around 25000 people didn't answer the salary question

In [45]:
df['Hobbyist']

Respondent
1        Yes
2         No
3        Yes
4         No
5        Yes
        ... 
88377    Yes
88601     No
88802     No
88816     No
88863    Yes
Name: Hobbyist, Length: 88883, dtype: object

In [46]:
#Find out how many times each distinct value have been answered
df['Hobbyist'].value_counts()

Hobbyist
Yes    71257
No     17626
Name: count, dtype: int64

In [47]:
df['SocialMedia']

Respondent
1          Twitter
2        Instagram
3           Reddit
4           Reddit
5         Facebook
           ...    
88377      YouTube
88601          NaN
88802          NaN
88816          NaN
88863     WhatsApp
Name: SocialMedia, Length: 88883, dtype: object

In [48]:
df['SocialMedia'].value_counts()

SocialMedia
Reddit                      14374
YouTube                     13830
WhatsApp                    13347
Facebook                    13178
Twitter                     11398
Instagram                    6261
I don't use social media     5554
LinkedIn                     4501
WeChat 微信                     667
Snapchat                      628
VK ВКонта́кте                 603
Weibo 新浪微博                     56
Youku Tudou 优酷                 21
Hello                          19
Name: count, dtype: int64

In [49]:
#lets find out what question was asked for SocialMedia column using our schema_df
schema_df.loc['SocialMedia']

QuestionText    What social media site do you use the most?
Name: SocialMedia, dtype: object

In [50]:
df['SocialMedia'].value_counts()

SocialMedia
Reddit                      14374
YouTube                     13830
WhatsApp                    13347
Facebook                    13178
Twitter                     11398
Instagram                    6261
I don't use social media     5554
LinkedIn                     4501
WeChat 微信                     667
Snapchat                      628
VK ВКонта́кте                 603
Weibo 新浪微博                     56
Youku Tudou 优酷                 21
Hello                          19
Name: count, dtype: int64

In [51]:
#lets find out in terms of proportion
df['SocialMedia'].value_counts(normalize= True)

SocialMedia
Reddit                      0.170233
YouTube                     0.163791
WhatsApp                    0.158071
Facebook                    0.156069
Twitter                     0.134988
Instagram                   0.074150
I don't use social media    0.065777
LinkedIn                    0.053306
WeChat 微信                   0.007899
Snapchat                    0.007437
VK ВКонта́кте               0.007141
Weibo 新浪微博                  0.000663
Youku Tudou 优酷              0.000249
Hello                       0.000225
Name: proportion, dtype: float64

In [52]:
#lets find out in terms of percentage
df['SocialMedia'].value_counts(normalize= True) *100

SocialMedia
Reddit                      17.023343
YouTube                     16.379076
WhatsApp                    15.807051
Facebook                    15.606902
Twitter                     13.498822
Instagram                    7.414996
I don't use social media     6.577685
LinkedIn                     5.330602
WeChat 微信                    0.789938
Snapchat                     0.743750
VK ВКонта́кте                0.714142
Weibo 新浪微博                   0.066322
Youku Tudou 优酷               0.024871
Hello                        0.022502
Name: proportion, dtype: float64

#### Group by: 
1. Split the object
2. Apply the function
3. Combine the result

In [53]:
df['Country'].value_counts()

Country
United States        20949
India                 9061
Germany               5866
United Kingdom        5737
Canada                3395
                     ...  
Tonga                    1
Timor-Leste              1
North Korea              1
Brunei Darussalam        1
Chad                     1
Name: count, Length: 179, dtype: int64

In [54]:
df.groupby(['Country'])

In [55]:
country_grp = df.groupby(['Country'])

In [56]:
country_grp.get_group('United States')

MainBranch Hobbyist  \
Respondent                                            
4           I am a developer by profession       No   
13          I am a developer by profession      Yes   
22          I am a developer by profession      Yes   
23          I am a developer by profession      Yes   
26          I am a developer by profession      Yes   
...                                    ...      ...   
78292                                  NaN       No   
82717                                  NaN       No   
83397                                  NaN      Yes   
85642                                  NaN       No   
88282                                  NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
4                                                       Never   
13          Less than once a month but more than once per ...   
22                                    Less than once per year   
23                                    Less than once per year   
26                                    Less than once per year   
...                                                       ...   
78292                              Once a month or more often   
82717                                 Less than once per year   
83397                                 Less than once per year   
85642                                 Less than once per year   
88282                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
4           The quality of OSS and closed source software ...   
13          OSS is, on average, of HIGHER quality than pro...   
22          OSS is, on average, of HIGHER quality than pro...   
23          The quality of OSS and closed source software ...   
26          The quality of OSS and closed source software ...   
...                                                       ...   
78292       OSS is, on average, of HIGHER quality than pro...   
82717       The quality of OSS and closed source software ...   
83397                                                     NaN   
85642       OSS is, on average, of LOWER quality than prop...   
88282       The quality of OSS and closed source software ...   

                                                   Employment        Country  \
Respondent                                                                     
4                                          Employed full-time  United States   
13                                         Employed full-time  United States   
22                                         Employed full-time  United States   
23                                         Employed full-time  United States   
26                                         Employed full-time  United States   
...                                                       ...            ...   
78292       Independent contractor, freelancer, or self-em...  United States   
82717                      Not employed, but looking for work  United States   
83397                      Not employed, but looking for work  United States   
85642       Independent contractor, freelancer, or self-em...  United States   
88282                      Not employed, but looking for work  United States   

           Student                                            EdLevel  \
Respondent                                                              
4               No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
13              No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
22              No  Some college/university study without earning ...   
23              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
26              No  Some college/university study without earning ...   
...            ...                                                ...   

There are 20949 people from USA

In [57]:
country_grp.get_group('India')

MainBranch Hobbyist  \
Respondent                                                    
8                      I code primarily as a hobby      Yes   
10                  I am a developer by profession      Yes   
15          I am a student who is learning to code      Yes   
50                  I am a developer by profession      Yes   
65                  I am a developer by profession      Yes   
...                                            ...      ...   
77339                                          NaN      Yes   
79795                                          NaN      Yes   
83862                                          NaN      Yes   
84299                                          NaN      Yes   
86012                                          NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
8                                     Less than once per year   
10                                 Once a month or more often   
15                                                      Never   
50                                 Once a month or more often   
65                                                      Never   
...                                                       ...   
77339                                 Less than once per year   
79795       Less than once a month but more than once per ...   
83862                              Once a month or more often   
84299                                                   Never   
86012                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
8           OSS is, on average, of HIGHER quality than pro...   
10          OSS is, on average, of HIGHER quality than pro...   
15          OSS is, on average, of HIGHER quality than pro...   
50          OSS is, on average, of LOWER quality than prop...   
65                                                        NaN   
...                                                       ...   
77339       OSS is, on average, of HIGHER quality than pro...   
79795       OSS is, on average, of HIGHER quality than pro...   
83862       OSS is, on average, of HIGHER quality than pro...   
84299       The quality of OSS and closed source software ...   
86012       OSS is, on average, of HIGHER quality than pro...   

                                    Employment Country         Student  \
Respondent                                                               
8           Not employed, but looking for work   India             NaN   
10                          Employed full-time   India              No   
15          Not employed, but looking for work   India  Yes, full-time   
50                          Employed full-time   India              No   
65                          Employed full-time   India              No   
...                                        ...     ...             ...   
77339                       Employed full-time   India  Yes, full-time   
79795                       Employed full-time   India              No   
83862                       Employed full-time   India  Yes, full-time   
84299                       Employed full-time   India  Yes, full-time   
86012                       Employed full-time   India              No   

                                                      EdLevel  \
Respondent                                                      
8                    Bachelor’s degree (BA, BS, B.Eng., etc.)   
10                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
15          Secondary school (e.g. American high school, G...   
50                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
65                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                       ...   
77339                Bachelor’s degree (BA, B

In [58]:
filt = (df['Country']== 'United States')
df.loc[filt]['SocialMedia'].value_counts()

SocialMedia
Reddit                      5700
Twitter                     3468
Facebook                    2844
YouTube                     2463
I don't use social media    1851
Instagram                   1652
LinkedIn                    1020
WhatsApp                     609
Snapchat                     326
WeChat 微信                     93
VK ВКонта́кте                  9
Weibo 新浪微博                     8
Hello                          2
Youku Tudou 优酷                 1
Name: count, dtype: int64

In [59]:
filt = (df['Country']== 'India')
df.loc[filt]['SocialMedia'].value_counts()

SocialMedia
WhatsApp                    2990
YouTube                     1820
LinkedIn                     955
Facebook                     841
Instagram                    822
Twitter                      542
Reddit                       473
I don't use social media     250
Snapchat                      23
Hello                          5
WeChat 微信                      5
VK ВКонта́кте                  4
Youku Tudou 优酷                 2
Weibo 新浪微博                     1
Name: count, dtype: int64

In [60]:
country_grp['SocialMedia'].value_counts().head(50)

Country              SocialMedia             
Afghanistan          Facebook                     15
                     YouTube                       9
                     I don't use social media      6
                     WhatsApp                      4
                     Instagram                     1
                     LinkedIn                      1
                     Twitter                       1
Albania              WhatsApp                     18
                     Facebook                     16
                     Instagram                    13
                     YouTube                      10
                     Twitter                       8
                     LinkedIn                      7
                     Reddit                        6
                     I don't use social media      4
                     Snapchat                      1
                     WeChat 微信                     1
Algeria              YouTube                      42


We grouped the socialMedia by Country

In [61]:
country_grp['SocialMedia'].value_counts().head()

Country      SocialMedia             
Afghanistan  Facebook                    15
             YouTube                      9
             I don't use social media     6
             WhatsApp                     4
             Instagram                    1
Name: count, dtype: int64

In [62]:
country_grp['SocialMedia'].value_counts().loc['India'] 

SocialMedia
WhatsApp                    2990
YouTube                     1820
LinkedIn                     955
Facebook                     841
Instagram                    822
Twitter                      542
Reddit                       473
I don't use social media     250
Snapchat                      23
Hello                          5
WeChat 微信                      5
VK ВКонта́кте                  4
Youku Tudou 优酷                 2
Weibo 新浪微博                     1
Name: count, dtype: int64

what is even the use of group by if we could find the same result by using the filt method above? because now we can find out social media of any country just by replacing the attribute in the .loc()

In [63]:
country_grp['SocialMedia'].value_counts().loc['United States']

SocialMedia
Reddit                      5700
Twitter                     3468
Facebook                    2844
YouTube                     2463
I don't use social media    1851
Instagram                   1652
LinkedIn                    1020
WhatsApp                     609
Snapchat                     326
WeChat 微信                     93
VK ВКонта́кте                  9
Weibo 新浪微博                     8
Hello                          2
Youku Tudou 优酷                 1
Name: count, dtype: int64

In [64]:
country_grp['SocialMedia'].value_counts(normalize = True).loc['Russian Federation']

SocialMedia
VK ВКонта́кте               0.303571
YouTube                     0.207418
I don't use social media    0.106456
Reddit                      0.096154
Twitter                     0.089286
Instagram                   0.064560
WhatsApp                    0.061813
Facebook                    0.046016
LinkedIn                    0.024038
Snapchat                    0.000687
Name: proportion, dtype: float64

In [65]:
country_grp['SalaryUSD'].median().loc['United States']

110000.0

In [66]:
#If we want to use multiple aggregate functions, we can pass a list
country_grp['SalaryUSD'].agg(['median','mean'])

median           mean
Country                                                      
Afghanistan                             6222.0  101953.333333
Albania                                10818.0   21833.700000
Algeria                                 7878.0   34924.047619
Andorra                               160931.0  160931.000000
Angola                                  7764.0    7764.000000
...                                        ...            ...
Venezuela, Bolivarian Republic of...    6384.0   14581.627907
Viet Nam                               11892.0   17233.436782
Yemen                                  11940.0   16909.166667
Zambia                                  5040.0   10075.375000
Zimbabwe                               19200.0   34046.666667

[179 rows x 2 columns]

In [67]:
#Median and mean salaries of specific countries
country_grp['SalaryUSD'].agg(['median','mean']).loc['United States']

median    110000.000000
mean      249546.254589
Name: United States, dtype: float64

In [68]:
filt = df['Country'] == 'India'
df.loc[filt]['LanguageWorkedWith'].str.contains('Python')

Respondent
8         True
10        True
15       False
50        True
65       False
         ...  
77339    False
79795     True
83862    False
84299     True
86012    False
Name: LanguageWorkedWith, Length: 9061, dtype: object

In [69]:
#find out the number of people who work with python
filt = df['Country'] == 'India'
df.loc[filt]['LanguageWorkedWith'].str.contains('Python').sum()

3105

In [70]:
country_grp['LanguageWorkedWith'].str.contains('Python').sum()

AttributeError: 'SeriesGroupBy' object has no attribute 'str'

In [71]:
country_grp['LanguageWorkedWith'].apply(lambda x: x.str.contains('Python').sum())

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

This doesn't tell us much, since we do not know out of how many people that answered the survey of these people know python, better way would be to find the percentage.

In [72]:
country_respondents = df['Country'].value_counts()
country_respondents

Country
United States        20949
India                 9061
Germany               5866
United Kingdom        5737
Canada                3395
                     ...  
Tonga                    1
Timor-Leste              1
North Korea              1
Brunei Darussalam        1
Chad                     1
Name: count, Length: 179, dtype: int64

In [73]:
country_uses_python = country_grp['LanguageWorkedWith'].apply(lambda x: x.str.contains('Python').sum())
country_uses_python

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

In [74]:
python_df = pd.concat([country_respondents, country_uses_python], axis = 'columns')
python_df

count  LanguageWorkedWith
Country                                     
United States      20949               10083
India               9061                3105
Germany             5866                2451
United Kingdom      5737                2384
Canada              3395                1558
...                  ...                 ...
Tonga                  1                   0
Timor-Leste            1                   1
North Korea            1                   0
Brunei Darussalam      1                   0
Chad                   1                   0

[179 rows x 2 columns]

we know how many people responded and how many of them know python

In [75]:
python_df.rename(columns = {'count' : 'Num_of_Respondents', 'LanguageWorkedWith':'KnowsPython'}, inplace = True)
python_df

Num_of_Respondents  KnowsPython
Country                                           
United States                   20949        10083
India                            9061         3105
Germany                          5866         2451
United Kingdom                   5737         2384
Canada                           3395         1558
...                               ...          ...
Tonga                               1            0
Timor-Leste                         1            1
North Korea                         1            0
Brunei Darussalam                   1            0
Chad                                1            0

[179 rows x 2 columns]

In [76]:
python_df['PctKnowsPython'] = python_df['KnowsPython']/python_df['Num_of_Respondents']*100
python_df

Num_of_Respondents  KnowsPython  PctKnowsPython
Country                                                           
United States                   20949        10083       48.131176
India                            9061         3105       34.267741
Germany                          5866         2451       41.783157
United Kingdom                   5737         2384       41.554820
Canada                           3395         1558       45.891016
...                               ...          ...             ...
Tonga                               1            0        0.000000
Timor-Leste                         1            1      100.000000
North Korea                         1            0        0.000000
Brunei Darussalam                   1            0        0.000000
Chad                                1            0        0.000000

[179 rows x 3 columns]

In [77]:
python_df.sort_values(by = 'PctKnowsPython', ascending = False, inplace = True)

In [78]:
python_df

Num_of_Respondents  KnowsPython  \
Country                                                             
Sao Tome and Principe                              1            1   
Timor-Leste                                        1            1   
Dominica                                           1            1   
Niger                                              1            1   
Turkmenistan                                       7            6   
...                                              ...          ...   
Cape Verde                                         3            0   
Lao People's Democratic Republic                   3            0   
Malawi                                             2            0   
Liberia                                            2            0   
Chad                                               1            0   

                                  PctKnowsPython  
Country                                           
Sao Tome and Principe                 100.000000  
Timor-Leste                           100.000000  
Dominica                              100.000000  
Niger                                 100.000000  
Turkmenistan                           85.714286  
...                                          ...  
Cape Verde                              0.000000  
Lao People's Democratic Republic        0.000000  
Malawi                                  0.000000  
Liberia                                 0.000000  
Chad                                    0.000000  

[179 rows x 3 columns]

In [79]:
python_df.head(50)

Num_of_Respondents  KnowsPython  \
Country                                                             
Sao Tome and Principe                              1            1   
Timor-Leste                                        1            1   
Dominica                                           1            1   
Niger                                              1            1   
Turkmenistan                                       7            6   
Mauritania                                         7            5   
Guinea                                             3            2   
Bahamas                                            3            2   
Guyana                                             3            2   
Uganda                                            72           47   
Iceland                                           49           30   
Haiti                                              5            3   
Benin                                              5            3   
Namibia                                            5            3   
Congo, Republic of the...                          7            4   
Oman                                              11            6   
Cuba                                              30           16   
Republic of Korea                                 39           20   
San Marino                                         2            1   
Sierra Leone                                       2            1   
Monaco                                             4            2   
South Korea                                      160           80   
Bhutan                                             2            1   
Seychelles                                         2            1   
Chile                                            206          102   
Finland                                          546          266   
Kenya                                            249          120   
United States                                  20949        10083   
Israel                                           952          457   
Ecuador                                           82           39   
Taiwan                                           187           88   
Montenegro                                        17            8   
Mongolia                                          17            8   
Switzerland                                      978          460   
United Republic of Tanzania                       32           15   
Hong Kong (S.A.R.)                               188           88   
Japan                                            391          182   
Maldives                                          13            6   
Canada                                          3395         1558   
Singapore                                        303          139   
New Zealand                                      524          236   
China                                            664          297   
Honduras                                          27           12   
France                                          2391         1054   
Iraq                                              39           17   
Thailand                                         214           92   
Democratic Republic of the Congo                   7            3   
Nepal                                            237          101   
Estonia                                          195           82   
Russian Federation                              1694          708   

                                  PctKnowsPython  
Country                                           
Sao Tome and Principe                 100.000000  
Timor-Leste                           100.000000  
Dominica                              100.000000  
Niger                                 100.000000  
Turkmenistan                           85.714286  
Mauritania                             71.428571  
Guinea                               

In [80]:
#Finding percetage by specific countries
python_df.loc['Japan']

Num_of_Respondents    391.000000
KnowsPython           182.000000
PctKnowsPython         46.547315
Name: Japan, dtype: float64

### Casting Datatypes and Handling Missing Values

lets go up and create na_vals list of Missing values

In [81]:
df['YearsCode'].head(10)

Respondent
1       4
2     NaN
3       3
4       3
5      16
6      13
7       6
8       8
9      12
10     12
Name: YearsCode, dtype: object

In [82]:
df['YearsCode'].mean()

TypeError: can only concatenate str (not "int") to str

In [83]:
df['YearsCode'] = df['YearsCode'].astype(float)

ValueError: could not convert string to float: 'Less than 1 year'

In [84]:
#all the unique value of the YearsCode column
df['YearsCode'].unique()

array(['4', nan, '3', '16', '13', '6', '8', '12', '2', '5', '17', '10',
       '14', '35', '7', 'Less than 1 year', '30', '9', '26', '40', '19',
       '15', '20', '28', '25', '1', '22', '11', '33', '50', '41', '18',
       '34', '24', '23', '42', '27', '21', '36', '32', '39', '38', '31',
       '37', 'More than 50 years', '29', '44', '45', '48', '46', '43',
       '47', '49'], dtype=object)

now lets replace these 2 values(less than 1 year and More than 50 years as numbers)

In [85]:
df['YearsCode'].replace('Less than 1 year', 0, inplace = True)

In [86]:
df['YearsCode'].replace('More than 50 years', 51, inplace = True)

In [87]:
df['YearsCode'] = df['YearsCode'].astype(float)

In [88]:
df['YearsCode'].mean()

11.662114216834588

In [89]:
round(df['YearsCode'].mean(),2)

11.66

In [90]:
df['YearsCode'].median()

9.0

### Reading/writing Data to Different Sources

In [91]:
filt = df['Country'] == 'India'
India_df = df.loc[filt] 
India_df

MainBranch Hobbyist  \
Respondent                                                    
8                      I code primarily as a hobby      Yes   
10                  I am a developer by profession      Yes   
15          I am a student who is learning to code      Yes   
50                  I am a developer by profession      Yes   
65                  I am a developer by profession      Yes   
...                                            ...      ...   
77339                                          NaN      Yes   
79795                                          NaN      Yes   
83862                                          NaN      Yes   
84299                                          NaN      Yes   
86012                                          NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
8                                     Less than once per year   
10                                 Once a month or more often   
15                                                      Never   
50                                 Once a month or more often   
65                                                      Never   
...                                                       ...   
77339                                 Less than once per year   
79795       Less than once a month but more than once per ...   
83862                              Once a month or more often   
84299                                                   Never   
86012                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
8           OSS is, on average, of HIGHER quality than pro...   
10          OSS is, on average, of HIGHER quality than pro...   
15          OSS is, on average, of HIGHER quality than pro...   
50          OSS is, on average, of LOWER quality than prop...   
65                                                        NaN   
...                                                       ...   
77339       OSS is, on average, of HIGHER quality than pro...   
79795       OSS is, on average, of HIGHER quality than pro...   
83862       OSS is, on average, of HIGHER quality than pro...   
84299       The quality of OSS and closed source software ...   
86012       OSS is, on average, of HIGHER quality than pro...   

                                    Employment Country         Student  \
Respondent                                                               
8           Not employed, but looking for work   India             NaN   
10                          Employed full-time   India              No   
15          Not employed, but looking for work   India  Yes, full-time   
50                          Employed full-time   India              No   
65                          Employed full-time   India              No   
...                                        ...     ...             ...   
77339                       Employed full-time   India  Yes, full-time   
79795                       Employed full-time   India              No   
83862                       Employed full-time   India  Yes, full-time   
84299                       Employed full-time   India  Yes, full-time   
86012                       Employed full-time   India              No   

                                                      EdLevel  \
Respondent                                                      
8                    Bachelor’s degree (BA, BS, B.Eng., etc.)   
10                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
15          Secondary school (e.g. American high school, G...   
50                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
65                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                       ...   
77339                Bachelor’s degree (BA, B

In [92]:
#send this india file to csv 
India_df.to_csv('data/modified.csv')

In [93]:
#tab seperated file
India_df.to_csv('data/modified.tav', sep = '\t')

In [96]:
#write to an excel file
India_df.to_excel('data/modified.xlsx')

In [97]:
test = pd.read_excel('data/modified.xlsx', index_col = 'Respondent')

In [98]:
test.head()

MainBranch Hobbyist  \
Respondent                                                    
8                      I code primarily as a hobby      Yes   
10                  I am a developer by profession      Yes   
15          I am a student who is learning to code      Yes   
50                  I am a developer by profession      Yes   
65                  I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
8              Less than once per year   
10          Once a month or more often   
15                               Never   
50          Once a month or more often   
65                               Never   

                                                   OpenSource  \
Respondent                                                      
8           OSS is, on average, of HIGHER quality than pro...   
10          OSS is, on average, of HIGHER quality than pro...   
15          OSS is, on average, of HIGHER quality than pro...   
50          OSS is, on average, of LOWER quality than prop...   
65                                                        NaN   

                                    Employment Country         Student  \
Respondent                                                               
8           Not employed, but looking for work   India             NaN   
10                          Employed full-time   India              No   
15          Not employed, but looking for work   India  Yes, full-time   
50                          Employed full-time   India              No   
65                          Employed full-time   India              No   

                                                      EdLevel  \
Respondent                                                      
8                    Bachelor’s degree (BA, BS, B.Eng., etc.)   
10                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
15          Secondary school (e.g. American high school, G...   
50                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
65                   Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
8           Computer science, computer engineering, or sof...   
10                                                        NaN   
15                                                        NaN   
50          Another engineering discipline (ex. civil, ele...   
65          Information systems, information technology, o...   

                                                     EduOther  \
Respondent                                                      
8           Taught yourself a new language, framework, or ...   
10                                                        NaN   
15          Taken an online course in programming or softw...   
50          Received on-the-job training in software devel...   
65                                                        NaN   

                             OrgSize  \
Respondent                             
8                                NaN   
10          10,000 or more employees   
15                               NaN   
50          10,000 or more employees   
65                20 to 99 employees   

                                                      DevType  YearsCode  \
Respondent                                                                 
8              Developer, back-end;Engineer, site reliability        8.0   
10          Data or business analyst;Data scientist or mac...       12.0   
15                                                    Student        3.0   
50                      Developer, back-end;DevOps specialist        7.0   
65                     Developer, front-end;Developer, mobile        2.0   

           Age1stCode YearsCodePro              CareerSat  \
Respondent                                                  
8                  16   

In [100]:
# Write to JSON file
India_df.to_json('data/modified.json')

In [101]:
#other way to write the json, the above was dictionary like(which is default), lets write it in list like
India_df.to_json('data/modified.json', orient = 'records', lines = True)

In [102]:
#Read json file
test = pd.read_json('data/modified.json', orient = 'records', lines = True)

In [103]:
test.head()

MainBranch Hobbyist  \
0             I code primarily as a hobby      Yes   
1          I am a developer by profession      Yes   
2  I am a student who is learning to code      Yes   
3          I am a developer by profession      Yes   
4          I am a developer by profession      Yes   

                  OpenSourcer  \
0     Less than once per year   
1  Once a month or more often   
2                       Never   
3  Once a month or more often   
4                       Never   

                                          OpenSource  \
0  OSS is, on average, of HIGHER quality than pro...   
1  OSS is, on average, of HIGHER quality than pro...   
2  OSS is, on average, of HIGHER quality than pro...   
3  OSS is, on average, of LOWER quality than prop...   
4                                               None   

                           Employment Country         Student  \
0  Not employed, but looking for work   India            None   
1                  Employed full-time   India              No   
2  Not employed, but looking for work   India  Yes, full-time   
3                  Employed full-time   India              No   
4                  Employed full-time   India              No   

                                             EdLevel  \
0           Bachelor’s degree (BA, BS, B.Eng., etc.)   
1        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
2  Secondary school (e.g. American high school, G...   
3           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                      UndergradMajor  \
0  Computer science, computer engineering, or sof...   
1                                               None   
2                                               None   
3  Another engineering discipline (ex. civil, ele...   
4  Information systems, information technology, o...   

                                            EduOther  \
0  Taught yourself a new language, framework, or ...   
1                                               None   
2  Taken an online course in programming or softw...   
3  Received on-the-job training in software devel...   
4                                               None   

                    OrgSize  \
0                      None   
1  10,000 or more employees   
2                      None   
3  10,000 or more employees   
4        20 to 99 employees   

                                             DevType  YearsCode Age1stCode  \
0     Developer, back-end;Engineer, site reliability        8.0         16   
1  Data or business analyst;Data scientist or mac...       12.0         20   
2                                            Student        3.0         13   
3              Developer, back-end;DevOps specialist        7.0         15   
4             Developer, front-end;Developer, mobile        2.0         17   

  YearsCodePro              CareerSat                 JobSat  \
0         None                   None                   None   
1           10  Slightly dissatisfied  Slightly dissatisfied   
2         None                   None                   None   
3            2     Slightly satisfied         Very satisfied   
4            2         Very satisfied         Very satisfied   

             MgrIdiot  MgrMoney   MgrWant  \
0                None      None      None   
1  Somewhat confident       Yes       Yes   
2                None      None      None   
3      Very confident  Not sure       Yes   
4      Very confident        No  Not sure   

                                             JobSeek          LastHireDate  \
0                                               None                  None   
1  I’m not actively looking, but I am open to new...         3-4 years ago   
2  I’m not actively looking, but I am open to new...  I've never had a job   
3  I’m not actively looking, but I am open to new...         1-2 years ago   
4  I’m not actively looking, but I am open to new...  Less than a 

In [ ]:
# Read database
import mysql.connector